crates/ra_analysis/src/imp.rs

```diff
@@ -305,6 +305,44 @@ impl AnalysisImpl {
             .collect()
     }
 
+    pub fn resolve_callable(&self, file_id: FileId, offset: TextUnit, token: &JobToken)
+        -> Option<(FnDescriptor, Option<usize>)> {
+
+        use ra_syntax::SyntaxKind;
+        use ra_syntax::ast::{ArgListOwner};
+        use ra_syntax::text_utils::{contains_offset_nonstrict};
+
+        let root = self.root(file_id);
+        let file = root.syntax(file_id);
+        let syntax = file.syntax();
+
+        // Find the calling expression
+        let call_expr: ast::CallExpr = find_node_at_offset(syntax, offset)?;
+        let call_offset = call_expr.syntax().range().start();
```

Comment:
I don't think this is quite right. If we're calling this function on something like `Foo::new(<|>)` then `call_offset` will point to the `F` and symbol will end up resolving to the `StructDef` of `Foo`. I think `call_offset` needs to be adjusted somehow.

Reply:
What do I do once I have the `PathExpr`? Do I have to walk the path somehow? It looks like there are two `NameRef`s in the `Foo::new()` example:

```
PATH_EXPR@[64; 70)
            PATH@[64; 70)
              PATH@[64; 65)
                PATH_SEGMENT@[64; 65)
                  NAME_REF@[64; 65)
                    IDENT@[64; 65) "F"
              COLONCOLON@[65; 67)
              PATH_SEGMENT@[67; 70)
                NAME_REF@[67; 70)
                  IDENT@[67; 70) "new"
```

Reply:
You'll need to go `PathExpr -> Path -> PathSegment -> NameRef`. I don't think AST has all relevant accessor functions at the moment, so you'll need to add them. In `grammar.ron`, 

change `"PathExpr": (),` to `"PathExpr": (options: ["Path"])`, and run `cargo gen-kinds`.